ClientRuntimes/CSharp/Microsoft.Rest.ClientRuntime/BatchDelegatingHandler.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Rest
+{
+    /// <summary>
+    /// HTTP batch handler.
+    /// It expects the caller to dispose HttpRequestMessage and HttpResponseMessage.
+    /// It assumes the web service is supporting batch requests as per this MSDN article:
+    /// https://blogs.msdn.microsoft.com/webdev/2013/11/01/introducing-batch-support-in-web-api-and-web-api-odata/
+    /// i.e. requests are batched into a single request inside the multipart content.
+    /// </summary>
+    public class BatchDelegatingHandler : DelegatingHandler
+    {
+        /// <summary>
+        /// HTTP method (e.g. POST or PUT) to use in issuing a batch call.
+        /// </summary>
+        private readonly HttpMethod batchMethod;
+
+        /// <summary>
+        /// URL to issue the batch call to.
+        /// </summary>
+        private readonly Uri batchURL;
+
+        /// <summary>
+        /// List of operations to issue in a batch.
+        /// </summary>
+        private List<HttpRequestMessage> requests;
+
+        /// <summary>
+        /// Array of HTTP content returned by the server in response to operations that are issued in a batch.
+        /// </summary>
+        private HttpContent[] responses;
+
+        /// <summary>
+        /// Event that tells all tasks that they can proceed.
+        /// </summary>
+        private ManualResetEvent batchCompleted;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="BatchDelegatingHandler"/> class.
+        /// </summary>
+        /// <param name="method">the type of method that the Batch operation is</param>
+        /// <param name="url">the URL that the Batch operation needs to hit</param>
+        public BatchDelegatingHandler(HttpMethod method, Uri url)
```

Comment:
Since the OData batch standard specifies that the request must be POST and it must be to the "[root]/$batch" endpoint, I think it would be best to make it impossible for users to do something different. The POST verb is easy enough to remove from the constructor arguments, but the batchURL is a little trickier (please see my [comment](https://github.com/Azure/autorest/pull/914/files#r63265593) in `IssueBatch()` in this file)